acme/acme/client.py

```diff
@@ -609,33 +620,42 @@ def poll_authorizations(self, orderr, deadline):
                 time.sleep(1)
         # If we didn't get a response for every authorization, we fell through
         # the bottom of the loop due to hitting the deadline.
-        if len(responses) > orderr.body.authorizations:
-            raise TimeoutError()
+        if len(responses) < len(orderr.body.authorizations):
+            raise errors.TimeoutError()
         failed = []
         for authzr in responses:
             if authzr.body.status != messages.STATUS_VALID:
                 for chall in authzr.body.challenges:
                     if chall.error != None:
                         failed.append(authzr)
         if len(failed) > 0:
-            raise ValidationError(failed)
+            raise errors.ValidationError(failed)
         return orderr.update(authorizations=responses)
 
     def finalize_order(self, orderr, deadline):
+        """Finalize an order and obtain a certificate.
+
+        :param messages.OrderResource orderr: order to finalize
+        :param datetime.datetime deadline: when to stop polling and timeout
+
+        :returns: finalized order
+        :rtype: messages.OrderResource
+
+        """
         csr = OpenSSL.crypto.load_certificate_request(
             OpenSSL.crypto.FILETYPE_PEM, orderr.csr_pem)
         wrapped_csr = messages.CertificateRequest(csr=jose.ComparableX509(csr))
-        self.net.post(latest.body.finalize, wrapped_csr)
+        self.net.post(orderr.body.finalize, wrapped_csr)
         while datetime.datetime.now() < deadline:
             time.sleep(1)
             response = self.net.get(orderr.uri)
             body = messages.Order.from_json(response.json())
             if body.error is not None:
-                raise IssuanceError(body.error)
+                raise errors.IssuanceError(body.error)
             if body.certificate is not None:
                 certificate_response = self.net.get(body.certificate).text
-                return orderr.update(fullchain_pem=certificate_response)
-        raise TimeoutError()
+                return orderr.update(body=body, fullchain_pem=certificate_response)
```

Comment:
Finalize doesn't change anything in `body`?

Reply:
Doesn't it? Earlier in the function, we post to the `finalize` endpoint and begin repeatedly making GET requests for the order until the `certificate` field is updated.

Reply:
Read this the other way around, thought you were removing body.